File: frontend/src/types/novel.ts
export interface Novel {
  id: string
  project_id: string
  title: string
  author?: string
  word_count: number
  format: string
  raw_text?: string
  cleaned_text?: string
  created_at: string
  updated_at: string
}

export interface Chapter {
  id: string
  novel_id: string
  chapter_number: number
  title?: string
  status: string
  paragraph_count?: number
  paragraph_stats?: ParagraphStats
  created_at: string
  updated_at?: string
}

export interface Paragraph {
  id: number | string
  text: string
  annotation?: string
  is_chapter_title?: boolean
}

export interface ParagraphStats {
  dialogue: number
  narration: number
  action: number
  description: number
  unmarked: number
}

// 编辑器段落
export interface EditorParagraph {
  id?: string
  paragraph_number: string
  text: string
  annotation_type?: AnnotationType
  annotation_content?: string
  comment?: string
  sort_order?: number
}

export type AnnotationType = 'dialogue' | 'narration' | 'action' | 'description'

export const ANNOTATION_CONFIG: Record<AnnotationType, { color: string; label: string; hex: string }> = {
  dialogue: { color: '#1677ff', label: '对话', hex: 'blue' },
  narration: { color: '#722ed1', label: '旁白', hex: 'purple' },
  action: { color: '#fa8c16', label: '动作', hex: 'orange' },
  description: { color: '#52c41a', label: '描述', hex: 'green' },
}

// 编辑器版本
export interface EditorVersion {
  id: string
  chapter_id: string
  version_number: number
  summary?: string
  created_at: string
}

export interface EditorVersionDetail extends EditorVersion {
  content_snapshot: EditorParagraph[]
}

// 版本历史
export interface VersionHistoryItem {
  id: string
  chapter_id: string
  version_number: number
  summary?: string
  created_at: string
}

// ===== 剧情拆解（Script）类型 =====

export interface ScriptShot {
  id: string
  shot_id: string
  content: string
  sort_order: number
}

export interface ScriptChapter {
  id: string
  title: string
  sort_order: number
  shots: ScriptShot[]
}

export interface ScriptResponse {
  chapters: ScriptChapter[]
}

export interface ScriptShotData {
  shotId: string
  content: string
}

export interface ScriptChapterData {
  chapterTitle: string
  shots: ScriptShotData[]
}

// ===== 分镜（Storyboard）类型 =====

export interface StoryboardShot {
  id: string
  shot_id: string
  storyboard_details: string
  sort_order: number
}

export interface StoryboardChapter {
  id: string
  title: string
  sort_order: number
  shots: StoryboardShot[]
}

export interface StoryboardResponse {
  chapters: StoryboardChapter[]
}

export interface StoryboardShotData {
  shotId: string
  storyboardDetails: string
}

export interface StoryboardChapterData {
  chapterTitle: string
  shots: StoryboardShotData[]
}

// ===== AI排版（Layout）类型 =====

export interface LayoutPage {
  id: string
  page_id: string
  layout_type: string
  page_purpose: string
  visual_focus: string
  shots: { shotId: string; content: string; storyboardDetails: string }[]
  sort_order: number
  image_prompt?: string
  image_url?: string
}

export interface LayoutChapter {
  id: string
  title: string
  sort_order: number
  pages: LayoutPage[]
}

export interface LayoutResponse {
  chapters: LayoutChapter[]
}

export interface LayoutPageData {
  pageId: string
  layoutType: string
  pagePurpose: string
  visualFocus: string
  shots: { shotId: string; content: string; storyboardDetails: string }[]
}

export interface LayoutChapterData {
  chapterTitle: string
  pages: LayoutPageData[]
}

File: backend/app/models/novel.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import Column, String, Integer, Text, DateTime, ForeignKey, JSON
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.core.database import Base
from app.core.base_model import VersionedMixin


class Project(Base):
    __tablename__ = "projects"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    name = Column(String(200), nullable=False)
    description = Column(Text, nullable=True)
    project_type = Column(String(20), default="comic")
    status = Column(String(20), default="draft")
    cover_image_url = Column(String(500), nullable=True)
    archived_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    novels = relationship("Novel", back_populates="project", cascade="all, delete-orphan")
    user = relationship("User", back_populates="projects")
    characters = relationship("Character", back_populates="project", cascade="all, delete-orphan")
    world_buildings = relationship("WorldBuilding", back_populates="project", cascade="all, delete-orphan")
    tasks = relationship("Task", back_populates="project", cascade="all, delete-orphan")
    character_relations = relationship("CharacterRelation", back_populates="project", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Project(id={self.id}, name='{self.name}', status='{self.status}')>"


class Novel(Base):
    __tablename__ = "novels"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id"), nullable=False, index=True)
    title = Column(String(200), nullable=False)
    author = Column(String(100), nullable=True)
    raw_text = Column(Text, nullable=False)
    cleaned_text = Column(Text, nullable=True)
    word_count = Column(Integer, default=0)
    format = Column(String(10), default="txt")
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    project = relationship("Project", back_populates="novels")
    chapters = relationship("Chapter", back_populates="novel", cascade="all, delete-orphan")
    script_chapters = relationship("ScriptChapter", back_populates="novel", cascade="all, delete-orphan")
    storyboard_chapters = relationship("StoryboardChapter", back_populates="novel", cascade="all, delete-orphan")
    layout_chapters = relationship("LayoutChapter", back_populates="novel", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Novel(id={self.id}, title='{self.title}', project_id={self.project_id})>"


class Chapter(Base):
    __tablename__ = "chapters"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    novel_id = Column(UUID(as_uuid=True), ForeignKey("novels.id"), nullable=False, index=True)
    chapter_number = Column(Integer, nullable=False)
    title = Column(String(200), nullable=True)
    content = Column(Text, nullable=True)
    status = Column(String(20), default="pending", index=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    novel = relationship("Novel", back_populates="chapters")
    paragraphs = relationship("Paragraph", back_populates="chapter", cascade="all, delete-orphan", order_by="Paragraph.paragraph_number")
    versions = relationship("EditorVersion", back_populates="chapter", cascade="all, delete-orphan", order_by="EditorVersion.version_number.desc()")

    def __repr__(self):
        return f"<Chapter(id={self.id}, number={self.chapter_number}, novel_id={self.novel_id})>"


class Paragraph(Base):
    __tablename__ = "paragraphs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    chapter_id = Column(UUID(as_uuid=True), ForeignKey("chapters.id"), nullable=False, index=True)
    paragraph_number = Column(String(10), nullable=False)  # 如 0001, 0001a, 0001b
    text = Column(Text, nullable=False, default="")
    annotation_type = Column(String(20), nullable=True)  # dialogue/narration/action/description
    annotation_content = Column(Text, nullable=True)     # 标注内容
    comment = Column(Text, nullable=True)                # 注释
    sort_order = Column(Integer, default=0)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    chapter = relationship("Chapter", back_populates="paragraphs")

    def __repr__(self):
        return f"<Paragraph(id={self.id}, num={self.paragraph_number}, chapter_id={self.chapter_id})>"


class EditorVersion(Base):
    __tablename__ = "editor_versions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    chapter_id = Column(UUID(as_uuid=True), ForeignKey("chapters.id"), nullable=False, index=True)
    version_number = Column(Integer, nullable=False)
    summary = Column(String(500), nullable=True)
    content_snapshot = Column(JSON, nullable=False)  # [{paragraph_number, text, annotation_type, annotation_content, comment}]
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    chapter = relationship("Chapter", back_populates="versions")

    def __repr__(self):
        return f"<EditorVersion(id={self.id}, chapter_id={self.chapter_id}, version={self.version_number})>"


class ScriptChapter(VersionedMixin, Base):
    """脚本章节 - 剧情拆解中的章节"""
    __tablename__ = "script_chapters"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    novel_id = Column(UUID(as_uuid=True), ForeignKey("novels.id"), nullable=False, index=True)
    title = Column(String(200), nullable=False)
    sort_order = Column(Integer, default=0)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    novel = relationship("Novel", back_populates="script_chapters")
    shots = relationship("ScriptShot", back_populates="chapter", cascade="all, delete-orphan", order_by="ScriptShot.sort_order")

    def __repr__(self):
        return f"<ScriptChapter(id={self.id}, title='{self.title}', novel_id={self.novel_id})>"


class ScriptShot(VersionedMixin, Base):
    """脚本镜头 - 剧情拆解中的单条镜头"""
    __tablename__ = "script_shots"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    chapter_id = Column(UUID(as_uuid=True), ForeignKey("script_chapters.id"), nullable=False, index=True)
    shot_id = Column(String(10), nullable=False)  # e.g., "01", "02"
    content = Column(Text, nullable=False, default="")
    sort_order = Column(Integer, default=0)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    chapter = relationship("ScriptChapter", back_populates="shots")

    def __repr__(self):
        return f"<ScriptShot(id={self.id}, shot_id='{self.shot_id}', chapter_id={self.chapter_id})>"
